Add how to configure streaming replication with custom WAL resource manager (#356)
pgroonga/pgroonga.github.io: ja/reference/custom-wal-resource-managers.md

Reviewer: `streaming-replication-wal-resource-manager.md`かな。

---
title: カスタムWALリソースマネージャー
---

# カスタムWALリソースマネージャー

> PGroongaは3.2.1からPostgreSQL組み込みのカスタムWALリソースマネージャー機能をサポートしています。

Reviewer: PostgreSQL 15以降が必須という情報も書いておいて。

これはPGroonga固有の手順です。

プライマリーとスタンバイでPGroongaをインストールします。

```bash
sudo apt install -y software-properties-common
sudo add-apt-repository -y universe
sudo add-apt-repository -y ppa:groonga/ppa
sudo apt install -y lsb-release
wget https://packages.groonga.org/ubuntu/groonga-apt-source-latest-$(lsb_release --codename --short).deb
sudo apt install -y -V ./groonga-apt-source-latest-$(lsb_release --codename --short).deb
sudo apt update
sudo apt install -y -V postgresql-16-pgdg-pgroonga
```

## [通常] プライマリーでストリーミングレプリケーション用にPostgreSQLを設定する {#configure-replication-primary}

これは通常の手順です。

プライマリーでだけ次のストリーミングレプリケーション用の設定を`postgresql.conf`に追加します。

  * `listen_addresses = 'localhost,192.168.0.30'`

多くのスタンバイがいる場合は`max_wal_senders`も設定する必要があります。`max_wal_senders`のデフォルト値は`10`です。スタンバイ2台では`10`で十分です。

[PostgreSQL：ドキュメント：レプリケーション][postgresql-replication]も参照してください。

`/etc/postgresql/16/main/postgresql.conf`:

変更前：

```vim
#listen_addresses = 'localhost'
```

変更後：

```vim
listen_addresses = 'localhost,192.168.0.30'
```

プライマリーでだけ以下のストリーミングレプリケーション用の設定を`pg_hba.conf`に追加します。

  * `192.168.0.0/24`からのレプリケーションユーザー`replicator`でのレプリケーション接続を許可します。

`/etc/postgresql/16/main/pg_hba.conf`:

変更前：

> ```vim

Reviewer: `vim`でいけるの！？

Author: 参考にした https://pgroonga.github.io/ja/reference/streaming-replication.html が `vim` になってたので、「`vim` の編集画面ってことかなー」と軽く流してましたが、`vim` script用のハイライトですね…。

local   replication     all                                     peer
host    replication     all             127.0.0.1/32            scram-sha-256
host    replication     all             ::1/128                 scram-sha-256
```

変更後：

```vim
local   replication     all                                     peer
host    replication     all             127.0.0.1/32            scram-sha-256
host    replication     all             ::1/128                 scram-sha-256
host    replication     all             192.168.0.0/24          scram-sha-256
```

プライマリーでだけレプリケーションユーザーを作成します。

```console
$ sudo -u postgres -H createuser --pwprompt --replication replicator
Enter password for new role: (passw0rd)
Enter it again: (passw0rd)
```

## [固有] プライマリーでPGroonga用にPostgreSQLを設定する {#configure-pgroonga-primary}

これはPGroonga固有の手順です。

> PGroongaのWALリソースマネージャー関連の設定とクラッシュセーフ関連の設定を追加する必要があります。

Reviewer: 「PGroonga WALリソースマネージャー」か「PGroongaのWALリソースマネージャー」かどっちかに統一しようか。

個人的には後者が好みかな。

> PGroonga WALリソースマネージャーようには
> [`pgroonga_wal_resource_manager`モジュール][pgroonga-wal-resource-manager]を
> [`shared_preload_libraries`パラメーター][postgresql-shared-preload-libraries]に追加して
> [`pgronga.enable_wal_resource_manager = on`][enable-wal-resource-manager]も追加する必要があります。

Reviewer: ```suggestion
PGroonga WALリソースマネージャー用には[`pgroonga_wal_resource_manager`モジュール][pgroonga-wal-resource-manager]を[`shared_preload_libraries`パラメーター][postgresql-shared-preload-libraries]に追加して[`pgronga.enable_wal_resource_manager = on`][enable-wal-resource-manager]も追加する必要があります。
```

```

変更後：

```vim
shared_preload_libraries = 'pgroonga_wal_resource_manager,pgroonga_crash_safer'
```

`/etc/postgresql/16/main/conf.d/pgroonga.conf`:

```vim
pgroonga.enable_wal_resource_manager = on
pgroonga.enable_crash_safe = on
```

`pgroonga_crash_safer`モジュールを使わない場合は、`shared_preload_libraries`から`pgroonga_crash_safer`を除いて、`pgroonga.enable_crash_safe = on`も消します。

この設定を反映するためにPostgreSQLを再起動します。

```bash
sudo -H systemctl restart postgresql
```

## [通常] プライマリーでデータを挿入する {#insert-primary}

これは通常の手順です。

プライマリーでだけ一般ユーザーを作成します。

```bash
sudo -u postgres -H createuser ${USER}
```

プライマリーでだけデータベースを作成します。

```bash
sudo -u postgres -H createdb --template template0 --locale C --encoding UTF-8 --owner ${USER} blog
```

プライマリーでだけ作成したデータベースにテーブルを追加します。

作成した`blog`データベースに接続します。

```bash
psql blog
```

`entries`テーブルを作成します。

```sql
CREATE TABLE entries (
  title text,
  body text
);
```

作成した`entries`テーブルにデータを追加します。

```sql
INSERT INTO entries VALUES ('PGroonga', 'PGroonga is a PostgreSQL extension for fast full text search that supports all languages. It will help us.');
INSERT INTO entries VALUES ('Groonga', 'Groonga is a full text search engine used by PGroonga. We did not know about it.');

> INSERT INTO entries VALUES ('PGroonga and replication', 'PGroonga 1.1.6 supports WAL based streaming replication. We should try it!');

Reviewer: ```suggestion
INSERT INTO entries VALUES ('PGroonga and replication', 'PGroonga 3.2.1 supports custom WAL resource manager. We should try it!');
```

```

作成したインデックスを確認します。

```sql
SET enable_seqscan TO off;
SELECT title FROM entries WHERE title &@~ 'replication';
--           title           
-- --------------------------
--  PGroonga and replication
-- (1 row)
```

## [固有] プライマリーでだけPGroonga関連のデータをフラッシュする {#flush-pgroonga-data-primary}

これはPGroonga固有の手順です。

プライマリーでだけメモリー上にあるPGroonga関連のデータを確実にディスクに書き出します。以下のどれかの方法を使います。

  1. `SELECT pgroonga_command('io_flush');`を実行する

  2. すべての接続を切断する

`pgroonga_command('io_flush')`を使う場合は次のようになります。

```sql
SELECT pgroonga_command('io_flush') AS command;
--                     command                    
-- -----------------------------------------------
--  [[0,1478446349.2241,0.1413860321044922],true]
-- (1 row)
```

プライマリーでは、次の`pg_basebackup`の手順が終わるまではPGroongaのインデックスを使っているテーブルを変更してはいけません。

## [通常] スタンバイで`pg_basebackup`を実行する {#pg-basebackup-standbys}

これは通常の手順です。

スタンバイでだけ`pg_basebackup`を実行します。`pg_basebackup`はプライマリーから現在のデータベースをコピーします。

スタンバイ：

```bash
sudo -H systemctl stop postgresql
sudo -u postgres -H rm -rf /var/lib/postgresql/16/main
```

スタンバイ1：

PGroonga WALリソースマネージャーを使う場合、[レプリケーションスロット][postgresql-replication-slots]のオプションも追加します。

* `--create-slot`

> * `--slot stanby1`

Reviewer: ```suggestion
* `--slot standby1`
```

158949/158949 kB (100%), 1/1 tablespace

```

## [固有] スタンバイでPGroonga用にPostgreSQLを設定する {#configure-pgroonga-standbys}

これはPGroonga固有の手順です。

次のモジュールを[`shared_preload_libraries`パラメーター][postgresql-shared-preload-libraries]に追加します。

  * [`pgroonga_wal_resource_manager`モジュール][pgroonga-wal-resource-manager]

注意: スタンバイでは`pgroonga_crash_safer`は必要ありません。[`pgroonga_wal_resource_manager`モジュール][pgroonga-wal-resource-manager]がリカバリーします。

スタンバイ：

`/etc/postgresql/16/main/postgresql.conf`:

変更前：

```vim
#shared_preload_libraries = ''
```

変更後：

```vim
shared_preload_libraries = 'pgroonga_wal_resource_manager'
```

スタンバイ：

`/etc/postgresql/16/main/conf.d/pgroonga.conf`:

```vim
pgroonga.enable_wal_resource_manager = on
```

## [通常] スタンバイでPostgreSQLを起動する {#start-standbys}

これは通常の手順です。

スタンバイでPostgreSQLを起動します。

```bash
sudo -H systemctl start postgresql
```

これで、プライマリーで挿入したデータをプライマリーで作成したPGroongaのインデックスで検索できます。

スタンバイ1：

```bash
psql blog
```

```sql
SET enable_seqscan TO off;
SELECT title FROM entries WHERE title &@~ 'replication';
--           title           
-- --------------------------
--  PGroonga and replication
-- (1 row)
```

`pg_basebacup`以降にプライマリーで追加したデータも検索できます。

プライマリー：

```sql

> INSERT INTO entries VALUES ('PostgreSQL 9.6 and replication', 'PostgreSQL supports generic WAL since 9.6. It is required for replication for PGroonga.');

Reviewer: ```suggestion
INSERT INTO entries VALUES ('PostgreSQL 9.6 and replication', 'PostgreSQL supports custom WAL resource manager since 15.');
```
